texecomConnect.py

```diff
@@ -508,286 +89,731 @@ def login(self):
+            self.areas[areaNumber] = Area(areaNumber)
+        return self.areas[areaNumber]
 
     def get_all_zones(self):
         for zoneNumber in range(1, self.numberOfZones + 1):
             zone = self.get_zone_details(zoneNumber)
-            self.zone[zoneNumber] = zone
+            self.zones[zoneNumber] = zone
+            if zone.zoneType != zone.ZONETYPE_UNUSED:
+                self.highestUsedZone = zoneNumber
+                self.associateZoneWithAreas(zone)
 
     def get_all_users(self):
-        panel_users = {12: 8, 24: 25, 48: 50, 64: 50, 88: 100, 168: 200, 640: 1000}
-        for usernumber in range(1, panel_users[self.numberOfZones]):
-            user = self.get_user(usernumber)
-            if user.valid():
-                self.user[usernumber] = user
-        user = User()
-        user.name = "Engineer"
-        self.user[0] = user
+        if self.numberOfUsers is not None:
+            for usernumber in range(1, self.numberOfUsers):
+                user = self.get_user(usernumber)
+                if user.valid():
+                    self.users[usernumber] = user
+            user = User()
+            user.name = "Engineer"
+            self.users[0] = user
 
     def get_all_areas(self):
-        panel_areas = {12: 2, 24: 2, 48: 4, 64: 4, 88: 8, 168: 16, 640: 64}
-        for areanumber in range(1, panel_areas[self.numberOfZones]):
+        for areanumber in range(1, self.numberOfAreas + 1):
             area = self.get_area_details(areanumber)
-            self.area[areanumber] = area
+            self.areas[areanumber] = area
+
+    def get_all_zones_state(self):
+        numZones = self.get_zone_state(1, self.highestUsedZone)
+        if numZones is not None and numZones != self.highestUsedZone:
+            self.log(
+                "get_all_zones_state request {:d} zones, got {:d}".format(
+                    self.highestUsedZone, numZones
+                )
+            )
+        return numZones
+
+    def get_changed_zones_state(self):
+        changedZonesBitmap = self.get_zone_changes()
+        if changedZonesBitmap is None:
+            return None
+        flags = int.from_bytes(changedZonesBitmap, "little")
+        zoneNum = 1
+        while zoneNum <= self.highestUsedZone:
+            if (flags & 1) == 0:
+                if flags == 0:
+                    break
+                flags = flags >> 1
+                zoneNum += 1
+            else:
+                startZoneNum = zoneNum
+                numZonesInBlock = 0
+                while (flags & 1) == 1 and numZonesInBlock < 168:
+                    flags = flags >> 1
+                    zoneNum += 1
+                    numZonesInBlock += 1
+                respNumZone = self.get_zone_state(startZoneNum, numZonesInBlock)
+                if respNumZone is None:
+                    return None
+        return True
+
+    def get_armed_area_state(self):
+        # we just track armed state (not part arming or part armed etc)
```

Comment:
I just noticed that this is overriding my part armed state.
When I part arm the alarm, it does to "armed_night" state for about a minute, then this polls for state, finds it armed (not part armed) and so updates the state to "armed_away". I guess that this is an additional flag to check?